src/System.IO.Compression/tests/DeflateStreamAsyncTests.cs

```diff
@@ -0,0 +1,166 @@
+        {
+            byte[] buffer = new byte[32];
+            string testFilePath = Path.Combine("GZTestData", "Canterbury", "GZcompressed", "alice29.txt" + ".gz");
+            using (var readStream = await ManualSyncMemoryStream.GetStreamFromFileAsync(testFilePath, false, true))
+            using (var unzip = new DeflateStream(readStream, CompressionMode.Decompress, true))
+            {
+                Task task = null;
+                try
+                {
+                    task = unzip.ReadAsync(buffer, 0, 32);
+                    Assert.Throws<InvalidOperationException>(() => { unzip.ReadAsync(buffer, 0, 32); }); // "overlapping read"
+                }
+                finally
+                {
+                   // clean up
+                    readStream.SetManualEvent();
+                    // internal assert that this should complete 
+                    Assert.True(task.Wait(100));
+                }
+            }
+        }
+
+        [Fact]
+        public static async Task OverlappingFlushAsync()
+        {
+            byte[] buffer = new byte[32];
+            string testFilePath = Path.Combine("GZTestData", "Canterbury", "GZcompressed", "alice29.txt" + ".gz");
+            using (var readStream = await ManualSyncMemoryStream.GetStreamFromFileAsync(testFilePath, false, true))
+            using (var unzip = new DeflateStream(readStream, CompressionMode.Decompress, true))
+            {
+                Task task = null;
+                try
+                {
+                    task = unzip.ReadAsync(buffer, 0, 32);
+                    Assert.Throws<InvalidOperationException>(() => { unzip.FlushAsync(); }); // "overlapping flush"
+                }
+                finally
+                {
+                    // clean up
+                    readStream.SetManualEvent();
+                    // internal assert that this should complete 
+                    Assert.True(task.Wait(100));
+                }
+            }
+        }
+
+        [Fact]
+        public static async Task OverlappingWriteAsync()
+        {
+            byte[] buffer = null;
+            string testFilePath = Path.Combine("GZTestData", "Canterbury", "alice29.txt");
+            using (var origStream = await LocalMemoryStream.readAppFileAsync(testFilePath))
+            {
+                buffer = origStream.ToArray();
+            }
+
+            using (var writeStream = new ManualSyncMemoryStream(false))
+            using (var zip = new DeflateStream(writeStream, CompressionMode.Compress))
+            {
+                Task task = null;
+                writeStream.ResetManualEvent();
+                try
+                {
+                    task = zip.WriteAsync(buffer, 0, 32);
+                    // TODO: following code doesn't work sometimes
+                    // Assert.Throws<InvalidOperationException>(() => { zip.WriteAsync(buffer, 32, 32); }, "overlapping write");
```

Comment:
FlushAsync is a bit different because we don't ever call the underlying stream's Flush method but instead call the underlying stream's write method via the FlushBuffers helper method.

If WriteAsync should be calling underlying.WriteAsync, then FlushAsync should be calling underlying.WriteAsync as well.

I'll open a new issue for this discussion and for now remove the OverlappingWriteAsync test.